app/actions/update_route_destinations.rb

```diff
@@ -126,6 +129,23 @@ def update_processes(processes_to_ports_map)
         raise Error.new(e.message)
       end
 
+      def compare_destination_hashes(new_route_mappings, existing_route_mappings)
+        new_route_mappings.reject do |new|
+          matching_route_mapping = existing_route_mappings.find do |existing|
+            new[:app_guid] == existing[:app_guid] &&
+              new[:process_type] == existing[:process_type] &&
+              new[:route_guid] == existing[:route_guid] &&
+              new[:app_port] == existing[:app_port]
+          end
+
+          if matching_route_mapping && matching_route_mapping[:protocol] != new[:protocol]
+            raise Error.new('Destination exists with conflicting protocol')
```

Comment:
Will users be able to hit this error when using the replace (`PATCH`) destinations endpoint?

If so, I don't think we want to error if the user is trying to replace all destinations, and one of them has a conflicting protocol with an existing destination.

Reply:
Currently, this method is only being used in the add (POST) cast so i think it's safe. However the name is misleading since it makes it seem more benign than it actually is. I could see someone trying to refactor this method into the update case and causing bugs down the line.

Reply:
While reviewing with Seth, we were also wondering if this error might be confusing in cases where the user hasn't specified a protocol in their request. This could then default to `http1` and conflict with an existing `http2` destination.

Reply:
> we were also wondering if this error might be confusing in cases where the user hasn't specified a protocol in their request. This could then default to http1 and conflict with an existing http2 destination.

Currently, an unspecified protocol in the request does default to `http1`, and that could conflict with an existing `http2` destination. @sethboyles is the ask to change the behavior in this case to not error and not update the protocol, or is it simply to change the error message?

Reply:
Hm, this is a tough one.  I think if we are keeping the null/default behavior, and telling api consumers it behaves like this: 

https://github.com/cloudfoundry/cloud_controller_ng/pull/2439/files#diff-ad56f3840f191aea04f2453955cda4fb69cd82a80272a01e17442a73fd40d239R28

then maybe we have to assume intentionality on the client's part and stick to the contract we state in the documentation.  That is, that `null` is an attempt to set a route destination's protocol to `http`.  It could otherwise lead to a surprising scenario wherein a user **thought** they set a destination to `http`, only for the API to silently ignore that 'insert', and who knows how long it would take to discover that their destination, which should be `http1`, is set to `http2`.

By erroring, we (perhaps annoyingly) alert the user that the route destination already exists with a different protocol.  But an immediate surprise is better than one that goes undiscovered, I think.  

However the current error message won't be very helpful if the client is attempting to insert several destinations, even for cases outside this particular null/default issue.  I think in the error message should include all identifying attributes of the offending destination(s), as well as the protocol (including what it was attempted to be set to and what the existing protocol is).

Reply:
We updated the error message to include more information: `Destination ROUTE_URI for app APP_GUID with process PROCESS_TYPE exists with conflicting protocol EXISTING_PROTOCOL, can't create destination with protocol NEW_PROTOCOL`